Remove unnecessary executor metrics

carterkozak opened the pull request:
*`executor.completed` is identical to the `count` value of the `executor.duration` timer. Dashboards will need to replace the former with the latter.
* `executor.scheduled.percent-of-period` is unused, our pipeline has filtered out the metric for a while now to reduce bloat.
* Scheduled executors no longer produce `executor.scheduled.once`, `executor.scheduled.repetitively`, or `executor.submitted` because the data isn't meaningful when load on the executor may be based on recurring tasks.
==COMMIT_MSG==
Remove unnecessary executor metrics. 
==COMMIT_MSG==

## Possible downsides?
This is a metrics _break_ which shouldn't be taken lightly.

carterkozak committed:
Remove unnecessary executor metrics

`executor.completed` is identical to the `count` value of
the `executor.duration` timer. Dashboards will need to
replace the former with the latter.

`executor.scheduled.percent-of-period` is unused, our pipeline has
filtered out the metric for a while now to reduce bloat.

`executor.scheduled.once` has been collapsed into `executor.submitted`
becuase they measure the same thing. This should allow more existing
dashboards to handle schedueld executors without updates.

carterkozak commented:
Adding do-not-merge until this has been thoroughly reviewed and considered.

carterkozak commented on tritium-metrics/src/main/metrics/metrics.yml:
```diff
@@ -64,10 +64,6 @@ namespaces:
         type: counter
         tags: [executor]
         docs: A gauge of the number of running tasks.
```
not a gauge
```suggestion
        docs: The number of running tasks.
```

carterkozak commented on tritium-metrics/src/main/metrics/metrics.yml:
```diff
@@ -76,10 +72,6 @@ namespaces:
-        type: meter
-        tags: [executor]
-        docs: A meter of the number of one-shot scheduled tasks. Applies only to scheduled executors.
       scheduled.repetitively:
         type: meter
         tags: [executor]
```
I'm not sure what this metric tells us, while we're breaking metrics, should this be removed? We could use `submitted` instead, or not record a metric in this case. Thoughts?

pkoenig10 replied:
If we're getting rid of `scheduled.once` we should also get rid of this metric. They're effectively the same metric just for one-shot vs repeated tasks.

pkoenig10 replied:
I've never really felt the need to distinguish between the two types of submissions (in my experience, one-shot scheduling is pretty rare) so I think just relying on `submitted` is sufficient.

carterkozak replied:
Might as well squash it into the `submitted` meter the same way as the `scheduled.once` metric I suppose?

carterkozak replied:
:+1:

schlosna replied:
This is marked by calls to `scheduleAtFixedRate` & `scheduleWithFixedDelay`.  I'm not sure how folks are making use of this, especially as a `meter` (could probably be a `counter` if we want to keep it). I guess one might want to track recurring task submission rates. `scheduleAtFixedRate` calls are wrapped with a `TaggedMetricsScheduledRunnable` that tracks `scheduledOverrun`, but we don't currently have anything here tracking tasks recurring tasks that have thrown and would not be rescheduled. My sense is the recurring tasks that have thrown and not been rescheduled might be of most value here.

@pkoenig10 may have thoughts from usage and when this was added in #231

schlosna replied:
ahh, actually looks like this was directly ported over for compatibility with Dropwizard metrics https://github.com/dropwizard/metrics/blob/cb3c59e4c71b4df761f5809d56e00dd0d82ef7cb/metrics-core/src/main/java/com/codahale/metrics/InstrumentedScheduledExecutorService.java#L92-L101

carterkozak committed:
Update tritium-metrics/src/main/metrics/metrics.yml

carterkozak committed:
Replace `executor.scheduled.repetitively` with `executor.submitted`

svc-changelog committed:
Add generated changelog entries

pkoenig10 commented on tritium-metrics/src/main/java/com/palantir/tritium/metrics/TaggedMetricsScheduledExecutorService.java:
```diff
@@ -35,52 +34,65 @@
+        // This means a wrapped same-thread executor will produce delayed 'submitted' values,
+        // however the results will work as expected for the more common cases in which
+        // either a queue is full, or the delegate has shut down.
+        submitted.mark();
```
This definitely isn't right.`submitted` should be updated when we submit the task to the executor - not when we schedule the tasks. Otherwise we'll be double counting submissions.

It's also weird to mark it here because it diverges from the meaning of `submitted` in non-scheduled executors.

carterkozak replied:
I'm not sure I follow, this matches how we mark `submitted` in non-scheduled executors. If I schedule a runnable to execute in 1 minute, the `submitted` metric works exactly like an instrumented fixed thread pool executor where the task sits in a queue for 1 minute.

> Otherwise we'll be double counting submissions.

This would cause tests to fail unless I've misunderstood your concern

pkoenig10 commented on tritium-metrics/src/main/java/com/palantir/tritium/metrics/TaggedMetricsScheduledExecutorService.java:
```diff
@@ -35,52 +34,65 @@
+        // This means a wrapped same-thread executor will produce delayed 'submitted' values,
+        // however the results will work as expected for the more common cases in which
+        // either a queue is full, or the delegate has shut down.
+        submitted.mark();
```
It's really confusing to use `submitted` for repetitive tasks. I typically view `submitted` as "how many tasks were submitted to the executor" and not "how many times was a executor method called".

carterkozak replied:
In this case a task is being submitted to the executor. It's not quite the same as a one-shot task in that it may be executed 0-N times rather than 0-1.
I could see an argument for excluding the metric entirely on recurring tasks, but I don't have a strong preference.

pkoenig10 replied:
Yeah I'd probably just exclude it entirely. I think the metric is significantly less useful when it's combined like this. I generally use `submitted` as a proxy for "how much work is being requested" and that doesn't really work when a single submitted task will be executed N times.

carterkozak replied:
Ya that's fair, at that point we might as well remove the submitted metric entirely from the scheduled executor because there's tricky interplay between recurring tasks and `execute`/`submit` which invalidates most of our assumptions anyhow.

pkoenig10 reviewed (commented):
Nevermind, I was

carterkozak committed:
Scheduled executors no longer produce  at all

carterkozak committed:
Remove 'executor.threads.terminated' metric

carterkozak commented:
I've updated to remove the `submitted` metric entirely from ScheduledExecutorService instrumentation because the usage patterns are such that the metric isn't very meaningful.

I've also removed the `executor.thread.finalized` meter because it's not terribly useful and can be estimated by subtracting the active counter from the created meter.

carterkozak committed:
Queued duration timer is only updated beyond a threshold

The queued duration estimate is only updated for tasks which
take longer than 250ms to begin executing. Not all tasks are
queued, so this should produce more realistic values in most
cases. However, it is possible that certain workloads do in
fact queue with very small durations, and we won't receive
that precise data anymore.

carterkozak commented:
Updated with another small optimization: The queuedDuration metric is only updated when tasks take longer than 250ms from the time they're submitted to the time they begin executing. Previously the queued duration metric was updated even when tasks were not queued at all, but with a relatively small value.

carterkozak commented:
Thanks, everyone!

svc-autorelease commented:
Released [0.30.0](https://github.com/palantir/tritium/releases/tag/0.30.0)
